From ebb8d8e643a6a3eee23058cc2ca237d38b8c9823 Mon Sep 17 00:00:00 2001
From: Arnd Bergmann <arnd@arndb.de>
Date: Thu, 4 Feb 2021 16:29:47 +0100
Subject: [PATCH] kallsyms: fix nonconverging kallsyms table with lld

commit efe6e3068067212b85c2d0474b5ee3b2d0c7adab upstream

ARM randconfig builds with lld sometimes show a build failure
from kallsyms:

  Inconsistent kallsyms data
  Try make KALLSYMS_EXTRA_PASS=1 as a workaround

The problem is the veneers/thunks getting added by the linker extend
the symbol table, which in turn leads to more veneers being needed,
so it may take a few extra iterations to converge.

This bug has been fixed multiple times before, but comes back every time
a new symbol name is used. lld uses a different set of identifiers from
ld.bfd, so the additional ones need to be added as well.

I looked through the sources and found that arm64 and mips define similar
prefixes, so I'm adding those as well, aside from the ones I observed. I'm
not sure about powerpc64, which seems to already be handled through a
section match, but if it comes back, the "__long_branch_" and "__plt_"
prefixes would have to get added as well.

Signed-off-by: Arnd Bergmann <arnd@arndb.de>
Signed-off-by: Masahiro Yamada <masahiroy@kernel.org>
Signed-off-by: Liu Haitao <haitao.liu@windriver.com>
---
 scripts/kallsyms.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/scripts/kallsyms.c b/scripts/kallsyms.c
index 5bf1ad013d36..b82cf6c9afb3 100644
--- a/scripts/kallsyms.c
+++ b/scripts/kallsyms.c
@@ -114,6 +114,13 @@ static bool is_ignored_symbol(const char *name, char type)
 	static const char * const ignored_prefixes[] = {
 		"__crc_",		/* modversions */
 		"__efistub_",		/* arm64 EFI stub namespace */
+		"__kvm_nvhe_",		/* arm64 non-VHE KVM namespace */
+		"__AArch64ADRPThunk_",	/* arm64 lld */
+		"__ARMV5PILongThunk_",	/* arm lld */
+		"__ARMV7PILongThunk_",
+		"__ThumbV7PILongThunk_",
+		"__LA25Thunk_",		/* mips lld */
+		"__microLA25Thunk_",
 		NULL
 	};
 
-- 
2.31.1

